### PR TITLE
DOC-7469: Improve governance around docs PRs

### PR DESCRIPTION
**Pull request by nickvigilante**
We as an organization have a contractual/legal obligation in some situations to ensure that our documentation is accurate. We want a member of the docs team to review each PR that comes in from a non-docs team member (this includes external contributors and CRL employees alike).

- If the PR is in draft, pass the check.
- Else:
  - If the author of the PR is a member of docs-prs, pass the check.
  - Else, check the reviewers of the PR.
    - If there is a review from at least one docs-prs member and the latest review from that member is approved, pass the check.
    - Else, fail the check.

Fixes DOC-7469

### CODE REVIEW

File: .github/scripts/docs-prs-reviews.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
+// Copyright 2023 The Cockroach Authors.
```
Comment (mdlinville) [inline .github/scripts/docs-prs-reviews.go:1]: fake news

**Review by mdlinville — APPROVED** [review]

**Review by mdlinville — CHANGES_REQUESTED** [review]
Needs more flair 🎉

**Review by mdlinville — CHANGES_REQUESTED** [review]
Nope

**Review by rail — CHANGES_REQUESTED** [review]
In overall it looks  great! But it's still Monday, so I'm going to be style pedantic. :)

File: .github/scripts/docs-prs-reviews.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	return parameters{
+		Token:    maybeEnv(githubTokenEnv, ""),
+		PrNumber: prNumber,
+		PrAuthor: maybeEnv(prAuthorEnv, ""),
+	}, nil
+}
+
+func maybeEnv(envKey, defaultValue string) string {
+	v := os.Getenv(envKey)
+	if v == "" {
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:145]: Ideally, none of these variables should ever be null. We need a GitHub token, a PR number, and a PR author every single time.

File: .github/scripts/docs-prs-reviews.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+	}
+	req, err := http.NewRequest("POST", graphQLURL, bytes.NewBuffer(requestBody))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", "token "+token)
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:251]: Done!

File: .github/scripts/docs-prs-reviews.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+func defaultEnvParameters() (parameters, error) {
+	const (
+		githubTokenEnv = "GITHUB_TOKEN"
+		prNumberEnv    = "PR_NUMBER"
+		prAuthorEnv    = "PR_AUTHOR"
+	)
+	prNumber, err := strconv.Atoi(maybeEnv(prNumberEnv, ""))
+	if err != nil {
+		fmt.Println(err)
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:133]: Done!

File: .github/scripts/docs-prs-reviews.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rail) [reply .github/scripts/docs-prs-reviews.go:145]: Totally understand. I usually use `os.Getenv()`, just wanted to mention and alternative.

File: .github/scripts/docs-prs-reviews.go [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+	return result, pageInfo.HasNextPage, pageInfo.EndCursor, nil
+}
+
+// queryGraphQL is the function that interfaces directly with the GitHub GraphQL API. Given a query, variables, and
+// token, it will return a struct containing the requested data or an error if one exists.
+func queryGraphQL(
+	query string, queryVariables map[string]interface{}, token string, out interface{},
+) error {
+	const graphQLURL = "https://api.github.com/graphql"
+	client := &http.Client{}
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:232]: Discussed with Rail offline, and we agreed this was optional. I'm marking this as resolved.

File: .github/scripts/docs-prs-reviews.go [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+	pageInfo := search.Data.Repository.PullRequest.Reviews.PageInfo
+	return result, pageInfo.HasNextPage, pageInfo.EndCursor, nil
+}
+
+// parseDocsPrReviewers takes a full list of reviews on a PR, takes only the PR reviews from members of docs-prs, and
+// checks if an approval has been made. Review comments (state = COMMENTED) have no effect on the review status.
+func parseDocsPrReviewers(reviews []review, docsPrsMembers []string) bool {
+	docsReviewers := make(map[string]prReviewParse)
+	for _, x := range reviews {
+		if stringInSlice(x.Author.Login, docsPrsMembers) {
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go]: Done!

File: .github/scripts/docs-prs-reviews.go [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+	docsOrganization = "cockroachdb"
+	docsRepo         = "docs"
+	docsPrsTeam      = "docs-prs"
+)
+
+func main() {
+	params, err := defaultEnvParameters()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error: %v\n", err)
+		os.Exit(1)
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go]: Ack. Thank you! Done!

File: .github/scripts/docs-prs-reviews.go [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+// searchDocsPrsTeamMembers passes in a GitHub API token and returns the list of docs-prs team members.
+func searchDocsPrsTeamMembers(token string) (map[string]interface{}, error) {
```
Comment (rail) [inline .github/scripts/docs-prs-reviews.go:156]:
```suggestion
func searchDocsPrsTeamMembers(token string) (map[string]struct{}, error) {
```

File: .github/scripts/docs-prs-reviews.go [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+	result := make(map[string]interface{})
```
Comment (rail) [inline .github/scripts/docs-prs-reviews.go:157]:
```suggestion
	result := make(map[string]struct{})
```

File: .github/scripts/docs-prs-reviews.go [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+	for _, member := range members {
+		result[member] = nil
+	}
+	for hasNextPage {
+		members, hasNextPage, nextCursor, err = searchDocsPrsTeamMembersSingle(nextCursor, token)
+		if err != nil {
+			return nil, fmt.Errorf("error searching docs-prs team members: %w", err)
+		}
+		for _, member := range members {
+			result[member] = nil
```
Comment (rail) [inline .github/scripts/docs-prs-reviews.go:171]:
```suggestion
			result[member] = struct{}{}
```

File: .github/scripts/docs-prs-reviews.go [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+// searchDocsPrsTeamMembers passes in a GitHub API token and returns the list of docs-prs team members.
+func searchDocsPrsTeamMembers(token string) (map[string]interface{}, error) {
+	result := make(map[string]interface{})
+	members, hasNextPage, nextCursor, err := searchDocsPrsTeamMembersSingle("", token)
+	if err != nil {
+		return nil, fmt.Errorf("error searching docs-prs team members: %w", err)
+	}
+	for _, member := range members {
+		result[member] = nil
```
Comment (rail) [inline .github/scripts/docs-prs-reviews.go:163]:
```suggestion
		result[member] = struct{}{}
```

File: .github/scripts/docs-prs-reviews.go [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+
+func maybeEnv(envKey, defaultValue string) string {
+	v := os.Getenv(envKey)
+	if v == "" {
+		return defaultValue
+	}
+	return v
+}
+
+func stringInSlice(a string, list []string) bool {
```
Comment (rail) [inline .github/scripts/docs-prs-reviews.go:146]: This is unused now. Kill it with fire :D

**Review by rail — CHANGES_REQUESTED** [review]
Sorry for the confusion with `interface{}`.  Almost there! :)

File: .github/scripts/docs-prs-reviews.go (hunk 8 above)
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:156]: Done!

File: .github/scripts/docs-prs-reviews.go (hunk 11 above)
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:163]: Done!

File: .github/scripts/docs-prs-reviews.go (hunk 12 above)
Comment (nickvigilante) [reply .github/scripts/docs-prs-reviews.go:146]: Done!

**Pull request merged**